test/functional/Java21Only/src/org/openj9/test/jep442/downcall/PrimitiveTypeTests1.java

```diff
@@ -263,9 +266,12 @@ public void test_strlenFromDefaultLibWithMemAddr_1() throws Throwable {
 		MemorySegment strlenSymbol = defaultLibLookup.find("strlen").get();
 		FunctionDescriptor fd = FunctionDescriptor.of(JAVA_LONG, ADDRESS);
 		MethodHandle mh = linker.downcallHandle(strlenSymbol, fd);
-		MemorySegment funcSegmt = arena.allocateUtf8String("JEP424 DOWNCALL TEST SUITES");
+		String str = "JEP424 DOWNCALL TEST SUITES";
+		MemorySegment funcSegmt = isZos
+				? arena.allocateArray(JAVA_BYTE, str.getBytes("Cp1047"))
+				: arena.allocateUtf8String(str);
 		long strLength = (long)mh.invoke(funcSegmt);
-		Assert.assertEquals(strLength, 27);
+		Assert.assertEquals(strLength, isZos ? 40 : 27);
```

Comment:
This is what happens on z/OS.  Based on what I observed with the output of  downcall via `printf` in native, the converted  string is followed by garbled data (which might be caused by the encoding mode on z/OS).

Reply:
To avoid counting the garbage data in the native string, one solution is to invoke `asSlice` on the segment.
```
String str = "JEP424 DOWNCALL TEST SUITES";
		MemorySegment funcSegmt = isZos
				? arena.allocateArray(JAVA_BYTE, str.getBytes("IBM1047").asSlice(0, str.length()));
```

Reply:
Unfortunately `asSlice` doesn't help as strlen still ends up with `40` instead of the string length in bytes (27).

Reply:
I double-checked the test code with arena.allocateArray(JAVA_BYTE, str.getBytes("UTF-8")).asSlice(0, 27) on zLinux which also ended up with `40`. That being said, something happens to `allocateArray` in which case it ends up with the same result whatever we can do to truncate the string.

Reply:
You don't want the `asSlice()` call.
Perhaps the issue is that `str.getBytes("IBM1047")` doesn't include the trailing `NUL` that `strlen()` will be looking for.

Reply:
Let me add "\0" to the end of string to see how it goes.

Reply:
With `String str = "JEP424 DOWNCALL TEST SUITES\0"`, `strLength = (long)mh.invoke(funcSegmt)` ends up with 27. I just updated these string by appending `\0` (the null terminator) to double-check the related tests.

Reply:
Just verified the tests on z/OS & zLinux with the latest changes which works good as expected.